### PR TITLE
OMG-211 transaction create endpoint

### PR DESCRIPTION
**Pull request by pnowosie**

### CODE REVIEW

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+       end}
+    end)
+  end
+
+  defp needed_funds(payments, %{currency: fee_currency, amount: fee_amount}) do
+    needed_funds =
+      payments
+      |> Enum.group_by(& &1.currency)
+      |> Enum.map(fn {k, v} ->
+        {k, v |> Enum.map(& &1.amount) |> Enum.sum()}
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: better name k, v

File: apps/omg_watcher/lib/web/validators/order.ex [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  @doc """
+  Parses and validates request body
+  """
+  @spec parse(map()) :: {:ok, UtxoSelection.order_t()} | {:error, any()}
+  def parse(params) do
+    with {:ok, owner} <- expect(params, "owner", :address),
+         {:ok, metadata} <- expect(params, "metadata", [:hash, :optional]),
+         {:ok, raw_payments} <- expect(params, "payments", :list),
+         {:ok, fee} <- parse_fee(Map.get(params, "fee", nil)),
```
Comment (pnowosie) [inline apps/omg_watcher/lib/web/validators/order.ex]: remove nil 
add test with missing "fee" in req

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        |> Enum.map(fn {_, utxos} -> length(utxos) end)
+        |> Enum.sum()
+
+      if utxo_count <= Transaction.max_inputs(),
+        do: create_transaction(funds, order) |> respond(:complete),
+        else: create_merge(owner, funds) |> respond(:intermediate)
+    end
+  end
+
+  defp select_utxo(utxos, needed_funds) do
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Comment data transformation here

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+         utxo ->
+           {0, [utxo]}
+       end}
+    end)
+  end
+
+  defp needed_funds(payments, %{currency: fee_currency, amount: fee_amount}) do
+    needed_funds =
+      payments
+      |> Enum.group_by(& &1.currency)
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Try Stream.*

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      end)
+      |> Map.new()
+
+    Map.update(needed_funds, fee_currency, fee_amount, &(&1 + fee_amount))
+  end
+
+  defp funds_sufficient?(utxo_selection) do
+    missing_funds =
+      utxo_selection
+      |> Enum.filter(fn {_, {short, _}} -> short > 0 end)
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Stream.*

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+  If also provided with receiver's address, creates and encodes a transaction.
+  """
+  @spec create_advice(%{Transaction.currency() => list(%DB.TxOutput{})}, order_t()) :: create_advice_t()
+  def create_advice(utxos, %{owner: owner, payments: payments, fee: fee} = order) do
+    needed_funds = needed_funds(payments, fee)
+    token_utxo_selection = select_utxo(utxos, needed_funds)
+
+    with {:ok, funds} <- funds_sufficient?(token_utxo_selection) do
+      utxo_count =
+        funds
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Stream.*

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    if Enum.empty?(missing_funds),
+      do: {:ok, utxo_selection |> Enum.map(fn {token, {_, utxos}} -> {token, utxos} end)},
+      else: {:error, :insufficient_funds, missing_funds}
+  end
+
+  defp create_transaction(utxos_per_token, %{owner: owner, payments: payments, metadata: metadata, fee: fee}) do
+    rests =
+      utxos_per_token
+      |> Enum.map(fn {token, utxos} ->
+        outputs = [fee | payments] |> Enum.filter(&(&1.currency == token)) |> Enum.map(& &1.amount) |> Enum.sum()
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Stream.*

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+          owner: Crypto.address_t(),
+          payments: nonempty_list(payment_t()),
+          fee: fee_t(),
+          metadata: binary() | nil
+        }
+
+  @type transaction_t() :: %{
+          inputs: nonempty_list(%DB.TxOutput{}),
+          outputs: nonempty_list(payment_t()),
+          fee: %{amount: integer, currency: Transaction.currency()},
```
Comment (kevsul) [inline apps/omg_watcher/lib/utxo_selection.ex]: Should this be `fee_t()` as defined above?

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+          txbytes: Transaction.tx_bytes() | nil,
+          metadata: Transaction.metadata()
+        }
+
+  @type create_advice_t() ::
```
Comment (pnowosie) [inline apps/omg_watcher/lib/utxo_selection.ex]: Just „advice-t”

File: apps/omg_watcher/lib/utxo_selection.ex (hunk 8 above)
Comment (pnowosie) [reply apps/omg_watcher/lib/utxo_selection.ex]: Yes, thank you!

File: apps/omg_watcher/lib/web/controllers/fallback.ex [hunk 10]
```diff
@@ -44,6 +44,14 @@ defmodule OMG.Watcher.Web.Controller.Fallback do
...
       code: "get_status:econnrefused",
       description: "Cannot connect to the Ethereum node."
+    },
+    insufficient_funds: %{
+      code: "transaction.create:insufficient_funds",
+      description: "Account balance is too low to satisfy the payment."
+    },
+    too_many_outputs: %{
+      code: "transaction.create:too_many_outputs",
+      description: "Effective number of outputs exceed allowed maximum."
```
Comment (kevsul) [inline apps/omg_watcher/lib/web/controllers/fallback.ex]: I think the word "Effective" is a bit confusing and unnecessary here.
```suggestion
      description: "Number of outputs exceed allowed maximum."
```

Comment (pnowosie) [reply apps/omg_watcher/lib/web/controllers/fallback.ex]: I wanted to express that we run out of available outputs trying to satisfy `payments ++ rests`, so this happens after request validation happened, at tx construction phase...

Comment (kevsul) [reply apps/omg_watcher/lib/web/controllers/fallback.ex]: Ok, I'm just not sure what the word 'Effective' means here. Is it that payments + fees + change exceed max outputs? Then maybe we can explicitly say that.
What I mean is that it should be obvious to the user what the problem is if they see this error.
And what should they do to fix it - remove a payment?

Comment (pnowosie) [reply apps/omg_watcher/lib/web/controllers/fallback.ex]: Agreed that the message would say: `Total number of payments + change + fees exceed maximum allowed transaction's outputs.`

File: apps/omg_watcher/test/web/controllers/transaction_test.exs [hunk 11]
```diff
@@ -428,4 +431,543 @@ defmodule OMG.Watcher.Web.Controller.TransactionTest do
...
+                   "payments" => [
+                     %{"amount" => payment, "currency" => @eth_hex, "owner" => Encoding.to_hex(alice.addr)}
+                   ],
+                   "fee" => %{"amount" => fee, "currency" => @eth_hex}
+                 }
+               )
+    end
+
+    @tag fixtures: [:alice, :more_utxos, :blocks_inserter]
+    test "does not return txbytes when spent owner is not provided", %{alice: alice} do
```
Comment (pik694) [inline apps/omg_watcher/test/web/controllers/transaction_test.exs]: _spend_ owner

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+  # the change). If this fails, we start to collect Utxos (starting from largest amount) which will cover the payment.
+  # We return {token, {change, [utxos for payment]}}, change > 0 means insufficient funds.
+  @spec select_utxo(%{Transaction.currency() => list(%DB.TxOutput{})}, %{Transaction.currency() => pos_integer()}) ::
+          list({Transaction.currency(), {integer, list(%DB.TxOutput{})}})
+  defp select_utxo(utxos, needed_funds) do
+    Enum.map(needed_funds, fn {token, need} ->
+      token_utxos = Map.get(utxos, token, [])
+
+      {token,
+       case Enum.find(token_utxos, fn %DB.TxOutput{amount: amount} -> amount == need end) do
```
[the remark this answers is not in the document]
Comment (pnowosie) [reply apps/omg_watcher/lib/utxo_selection.ex]: As discussed: selection algorithm can change, we don't want to introduce constrains regarding data here, yet.

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+           end)
+
+         utxo ->
+           {0, [utxo]}
+       end}
+    end)
+  end
+
+  # Sums up payments by token. Fee is included.
+  defp needed_funds(payments, %{currency: fee_currency, amount: fee_amount}) do
```
[the remark this answers is not in the document]
Comment (pnowosie) [reply apps/omg_watcher/lib/utxo_selection.ex]: That's correct. One needs to cover fee entirely in one of fee-supported currencies.

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+  @spec select_utxo(%{Transaction.currency() => list(%DB.TxOutput{})}, %{Transaction.currency() => pos_integer()}) ::
+          list({Transaction.currency(), {integer, list(%DB.TxOutput{})}})
+  defp select_utxo(utxos, needed_funds) do
+    Enum.map(needed_funds, fn {token, need} ->
+      token_utxos = Map.get(utxos, token, [])
+
+      {token,
+       case Enum.find(token_utxos, fn %DB.TxOutput{amount: amount} -> amount == need end) do
+         nil ->
+           Enum.reduce_while(token_utxos, {need, []}, fn
```
[the remark this answers is not in the document]
Comment (pnowosie) [reply apps/omg_watcher/lib/utxo_selection.ex]: That's true. Ehh, I'll make a not in a comment before function.

File: apps/omg_watcher/lib/utxo_selection.ex [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+      end)
+      |> Map.new()
+
+    Map.update(needed_funds, fee_currency, fee_amount, &(&1 + fee_amount))
+  end
+
+  defp funds_sufficient?(utxo_selection) do
+    missing_funds =
+      utxo_selection
+      |> Stream.filter(fn {_, {change, _}} -> change > 0 end)
```
[the remark this answers is not in the document]
Comment (pnowosie) [reply apps/omg_watcher/lib/utxo_selection.ex]: Agree. Here better name is `missing`

**Pull request merged**